src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java

```diff
@@ -0,0 +1,196 @@
+        singleKeyStoreProvider.put(dummyProviderName1, new DummyKeyStoreProvider());
+
+        multipleKeyStoreProviders.put(dummyProviderName2, new DummyKeyStoreProvider());
+        multipleKeyStoreProviders.put(dummyProviderName3, new DummyKeyStoreProvider());
+        multipleKeyStoreProviders.put(dummyProviderName4, new DummyKeyStoreProvider());
+
+        if (null != SQLServerConnection.globalCustomColumnEncryptionKeyStoreProviders) {
+            globalKeyStoreProviders = new HashMap<String, SQLServerColumnEncryptionKeyStoreProvider>(SQLServerConnection.globalCustomColumnEncryptionKeyStoreProviders);
+        }
+        
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+    }
+
+    @Test
+    public void testNullProviderMap() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = null;
+        assertExceptionIsThrownForAllCustomProviderCaches(providers, TestUtils.formatErrorMsg("R_CustomKeyStoreProviderMapNull"));
+    }
+
+    @Test
+    public void testInvalidProviderNameOnGlobal() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put("MSSQL_DUMMY", new DummyKeyStoreProvider());
+
+        assertExceptionIsThrownForGlobalCustomProviderCache(providers, TestUtils.formatErrorMsg("R_InvalidCustomKeyStoreProviderName"));
+    }
+
+    @Test
+    public void testInvalidProviderNameAllLevels() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put(WINDOWS_KEY_STORE_NAME, new DummyKeyStoreProvider());
+        assertExceptionIsThrownForAllCustomProviderCaches(providers, TestUtils.formatErrorMsg("R_InvalidCustomKeyStoreProviderName"));
+    }
+
+    @Test
+    public void testNullProviderAllLevels() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put("DUMMY_PROVIDER", null);
+        assertExceptionIsThrownForAllCustomProviderCaches(providers, TestUtils.formatErrorMsg("R_CustomKeyStoreProviderValueNull"));
+
+    }
+
+    @Test
+    public void testEmptyProviderNameAllLevels() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put(" ", new DummyKeyStoreProvider());
+        assertExceptionIsThrownForAllCustomProviderCaches(providers, TestUtils.formatErrorMsg("R_EmptyCustomKeyStoreProviderName"));
+    }
+
+    @Test
+    public void testSetProviderOnlyOnceOnGlobal() {
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put("DUMMY_PROVIDER", new DummyKeyStoreProvider());
+
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+
+        try {
+            SQLServerConnection.registerColumnEncryptionKeyStoreProviders(providers);
+        }
+        catch (Exception e) {
+            fail(TestResource.getResource("R_unexpectedException"));

```

Comment:
can you append e.getMessage() to this like in other places? That makes it easier to debug failures. (multiple places)

Reply:
changed